servers/rendering/rendering_device_driver.h

```diff
@@ -0,0 +1,619 @@
+
+	// ----- CACHE -----
+
+	virtual bool pipeline_cache_create(const Vector<uint8_t> &p_data) = 0;
+	virtual void pipeline_cache_free() = 0;
+	virtual size_t pipeline_cache_query_size() = 0;
+	virtual Vector<Vector<uint8_t>> pipeline_cache_serialize() = 0;
+
+	/*******************/
+	/**** RENDERING ****/
+	/*******************/
+
+	// ----- SUBPASS -----
+
+	enum AttachmentLoadOp {
+		ATTACHMENT_LOAD_OP_LOAD = 0,
+		ATTACHMENT_LOAD_OP_CLEAR = 1,
+		ATTACHMENT_LOAD_OP_DONT_CARE = 2,
+		ATTACHMENT_LOAD_OP_NONE = 1000400000,
+	};
+
+	enum AttachmentStoreOp {
+		ATTACHMENT_STORE_OP_STORE = 0,
+		ATTACHMENT_STORE_OP_DONT_CARE = 1,
+		ATTACHMENT_STORE_OP_NONE = 1000301000,
+	};
+
+	struct Attachment {
+		DataFormat format = DATA_FORMAT_MAX;
+		TextureSamples samples = TEXTURE_SAMPLES_MAX;
+		AttachmentLoadOp load_op = ATTACHMENT_LOAD_OP_LOAD;
+		AttachmentStoreOp store_op = ATTACHMENT_STORE_OP_STORE;
+		AttachmentLoadOp stencil_load_op = ATTACHMENT_LOAD_OP_LOAD;
+		AttachmentStoreOp stencil_store_op = ATTACHMENT_STORE_OP_STORE;
+		TextureLayout initial_layout = TEXTURE_LAYOUT_UNDEFINED;
+		TextureLayout final_layout = TEXTURE_LAYOUT_UNDEFINED;
+	};
+
+	struct AttachmentReference {
+		static const uint32_t UNUSED = 0xffffffff;
+		uint32_t attachment = UNUSED;
+		TextureLayout layout = TEXTURE_LAYOUT_UNDEFINED;
+		BitField<TextureAspectBits> aspect;
+	};
+
+	struct Subpass {
+		LocalVector<AttachmentReference> input_references;
+		LocalVector<AttachmentReference> color_references;
+		AttachmentReference depth_stencil_reference;
+		LocalVector<AttachmentReference> resolve_references;
+		LocalVector<uint32_t> preserve_attachments;
+		AttachmentReference vrs_reference;
+	};
+
+	struct SubpassDependency {
+		uint32_t src_subpass = 0xffffffff;
+		uint32_t dst_subpass = 0xffffffff;
+		BitField<PipelineStageBits> src_stages;
+		BitField<PipelineStageBits> dst_stages;
+		BitField<BarrierAccessBits> src_access;
+		BitField<BarrierAccessBits> dst_access;
+	};
+
+	virtual RenderPassID render_pass_create(VectorView<Attachment> p_attachments, VectorView<Subpass> p_subpasses, VectorView<SubpassDependency> p_subpass_dependencies, uint32_t p_view_count) = 0;
+	virtual void render_pass_free(RenderPassID p_render_pass) = 0;
+
+	// ----- COMMANDS -----
+
+	union RenderPassClearValue {
+		Color color = {};
+		struct {
+			float depth;
+			uint32_t stencil;
+		};
+	};
+
+	struct AttachmentClear {
+		BitField<TextureAspectBits> aspect;
+		uint32_t color_attachment = 0xffffffff;
+		RenderPassClearValue value;
+	};
+
+	struct AttachmentClearRect {
+		Rect2i rect;
+		uint32_t base_layer = 0xffffffff;
+		uint32_t layer_count = 0xffffffff;
+	};
+
+	virtual void command_begin_render_pass(CommandBufferID p_cmd_buffer, RenderPassID p_render_pass, FramebufferID p_framebuffer, CommandBufferType p_cmd_buffer_type, const Rect2i &p_rect, VectorView<RenderPassClearValue> p_clear_values) = 0;
```

Comment:
Minor nitpick, what's the benefit of CommandBufferType being specified here? Shouldn't we be able to derive it from the command buffer itself? Is it a validation measure?

Reply:
Yeah thats a good point. That said I think we may want to simplify this.

Reply:
Well, since `RD` knows if it's about a primary or secondary command buffer, that argument avoids the need for `RDD` to track it and be able to map `RD::CommandBufferID` to `VkCommandBuffer` directly. The parameter is leveraged this way:
```
	VkSubpassContents vk_subpass_contents = p_cmd_buffer_type == COMMAND_BUFFER_TYPE_PRIMARY ? VK_SUBPASS_CONTENTS_INLINE : VK_SUBPASS_CONTENTS_SECONDARY_COMMAND_BUFFERS;
	vkCmdBeginRenderPass((VkCommandBuffer)p_cmd_buffer.id, &render_pass_begin, vk_subpass_contents);
```

I'm not sure how orthogonal are command buffer level and subpass content level as concepts, but the fact is that currently `RD` uses them 1:1, so I just took advantage.